src/main/java/io/aiven/kafka/auth/AivenAclAuthorizerV2.java

```diff
@@ -0,0 +1,224 @@
+    }
+
+    @Override
+    public void configure(final java.util.Map<String, ?> configs) {
+        final AivenAclAuthorizerConfig config = new AivenAclAuthorizerConfig(configs);
+
+        auditor = config.getAuditor();
+        logDenials = config.logDenials();
+
+        configFile = config.getConfigFile();
+        final AclJsonReader jsonReader = new AclJsonReader(configFile.toPath());
+        cacheReference.set(VerdictCache.create(loadAcls(jsonReader)));
+        final AtomicReference<WatchKey> watchKeyReference = new AtomicReference<>(subscribeToAclChanges(configFile));
+        scheduledExecutorService.scheduleWithFixedDelay(() -> {
+            final WatchKey watchKey = watchKeyReference.get();
+            if (watchKey != null) {
+                final List<WatchEvent<?>> watchEvents = watchKey.pollEvents();
+
+                watchEvents.stream().filter(watchEvent -> {
+                    @SuppressWarnings("unchecked")
+                    final Path path = ((WatchEvent<Path>) watchEvent).context();
+                    return path.endsWith(configFile.getName());
+                }).findFirst().ifPresent(watchEvent -> {
+                    LOGGER.info("{}: {}, Modified: {}",
+                            watchEvent.kind(), watchEvent.context(), configFile.lastModified());
+                    cacheReference.set(VerdictCache.create(loadAcls(jsonReader)));
+                });
+                if (!watchKey.reset()) {
+                    watchKeyReference.compareAndSet(watchKey, null);
+                }
+            } else {
+                watchKeyReference.set(subscribeToAclChanges(configFile));
+            }
+        }, 0, config.configRefreshInterval(), TimeUnit.MILLISECONDS);
+    }
+
+    private WatchKey subscribeToAclChanges(final File configFile) {
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+            return configFile.toPath().toAbsolutePath().getParent()
+                    .register(watchService, StandardWatchEventKinds.ENTRY_MODIFY,
+                            StandardWatchEventKinds.ENTRY_CREATE, StandardWatchEventKinds.ENTRY_DELETE);
+        } catch (final IOException e) {
+            LOGGER.error("Failed to subscribe to ACL configuration changes", e);
+            return null;
+        }
+    }
+
+    @Override
+    public void close() {
+        auditor.stop();
+        scheduledExecutorService.shutdownNow();
+        try {
+            watchService.close();
+        } catch (final IOException e) {
+            LOGGER.error("Failed to stop watch service", e);
+        }
+    }
+
+    @Override
+    public boolean authorize(final Session session,
+                             final Operation operation,
+                             final Resource resource) {
+        final KafkaPrincipal principal =
+            Objects.nonNull(session.principal())
+                ? session.principal()
+                : KafkaPrincipal.ANONYMOUS;
```

Comment:
No need to change, but I just like this:
```suggestion
        final KafkaPrincipal principal = Objects.requireNonNullElse(session.principal(), KafkaPrincipal.ANONYMOUS);
```